src/coreclr/jit/compiler.cpp

```diff
@@ -9693,3 +9712,311 @@ const char* Compiler::devirtualizationDetailToString(CORINFO_DEVIRTUALIZATION_DE
+
+                case AddressExposedReason::TOO_CONSERVATIVE:
+                    m_tooConservative++;
+                    break;
+
+                case AddressExposedReason::ESCAPE_ADDRESS:
+                    m_escapeAddress++;
+                    break;
+
+                case AddressExposedReason::WIDE_INDIR:
+                    m_wideIndir++;
+                    break;
+
+                case AddressExposedReason::OSR_EXPOSED:
+                    m_osrExposed++;
+                    break;
+
+                case AddressExposedReason::STRESS_LCL_FLD:
+                    m_stressLclFld++;
+                    break;
+
+                case AddressExposedReason::COPY_FLD_BY_FLD:
+                    m_copyFldByFld++;
+                    break;
+
+                case AddressExposedReason::DISPATCH_RET_BUF:
+                    m_dispatchRetBuf++;
+                    break;
+
+                default:
+                    unreached();
+                    break;
+            }
+        }
+    }
+}
+
+void Compiler::EnregStructStat::Dump(FILE* fout) const
+{
+    const unsigned totalNumberOfNotStructVars =
+        s_enregStructStat.m_totalNumberOfVars - s_enregStructStat.m_totalNumberOfStrutVars;
+    const unsigned totalNumberOfNotStructEnregVars =
+        s_enregStructStat.m_totalNumberOfEnregVars - s_enregStructStat.m_totalNumberOfStructEnregVars;
+    const unsigned notEnreg = s_enregStructStat.m_totalNumberOfVars - s_enregStructStat.m_totalNumberOfEnregVars;
+
+    fprintf(fout, "\nLocals enregistration statistics:\n");
+    if (m_totalNumberOfVars == 0)
+    {
+        fprintf(fout, "No locals to report.\n");
+        return;
+    }
+    fprintf(fout, "total number of locals: %d, number of enregistered: %d, notEnreg: %d, ratio: %.2f\n",
+            m_totalNumberOfVars, m_totalNumberOfEnregVars, m_totalNumberOfVars - m_totalNumberOfEnregVars,
+            (float)m_totalNumberOfEnregVars / m_totalNumberOfVars);
+
+    if (m_totalNumberOfStrutVars != 0)
+    {
+        fprintf(fout, "total number of struct locals: %d, number of enregistered: %d, notEnreg: %d, ratio: %.2f\n",
+                m_totalNumberOfStrutVars, m_totalNumberOfStructEnregVars,
+                m_totalNumberOfStrutVars - m_totalNumberOfStructEnregVars,
+                (float)m_totalNumberOfStructEnregVars / m_totalNumberOfStrutVars);
+    }
+
+    const unsigned numberOfPrimitiveLocals = totalNumberOfNotStructVars - totalNumberOfNotStructEnregVars;
+    if (numberOfPrimitiveLocals != 0)
+    {
+        fprintf(fout, "total number of primitive locals: %d, number of enregistered: %d, notEnreg: %d, ratio: %.2f\n",
+                totalNumberOfNotStructVars, totalNumberOfNotStructEnregVars, numberOfPrimitiveLocals,
+                (float)totalNumberOfNotStructEnregVars / totalNumberOfNotStructVars);
+    }
+
+    if (notEnreg == 0)
+    {
+        fprintf(fout, "All locals are enregistered.\n");
+        return;
+    }
+
+    if (m_addrExposed != 0)
```

Comment:
It will be better readable if we have a macro something like this:

```c++
#define PRINT_STATS(stat, total) \
    if (stat != 0) \
    { \
        fprintf(fout, #stat "%d, ratio: %.2f\n", stat, (float)stat / total); \
    } \

PRINT_STATS(m_addrExposed, notEnreg)
PRINT_STATS(m_notRegSizeStruct , notEnreg)
...
PRINT_STATS(m_copyFldByFld, m_addrExposed)
```

Reply:
ok, done.